### ISSUE TITLE
Shortcuts for Geany Preview

### ISSUE THREAD

**Issue by ralf3u** [xiota/geanylua-scripts#7]
Shortcut names
1: `Toggle between only editor and editor/sidebar`
2: `Toggle between only editor and editor/sidebar; focus stays in the editor`
3: `Toggle between only sidebar and editor/sidebar`
4: `Toggle between only editor and only sidebar`
5: `Exit "Toggle between only editor and only sidebar"`

Shortcuts in action
1: if "only editor" is not shown, then (show "only editor" & put editor in focus), else (show editor/sidebar & put sidebar in focus)
2: if "only editor" is not shown, then (show "only editor" & put editor in focus), else (show editor/sidebar & put editor in focus)
3: if "only sidebar" is not shown, then (show "only sidebar" & put sidebar in focus), else (show editor/sidebar & put editor in focus)
4: if the shortcuts 1, 2 and 3 are blocked, then (if "only editor" is shown, then (show "only sidebar" & put sidebar in focus), else (show "only editor" & put editor in focus)); if editor/sidebar is shown, then (block the shortcuts 1, 2 and 3 & show "only editor" & put editor in focus)
5: if the shortcuts 1, 2 and 3 are blocked, then (unblock the shortcuts 1, 2 and 3 & show editor/sidebar & put editor in focus)

Explication
Toggling the shortcut 1 or toggling the shortcut 2 is interesting, if one don't want to see the whole result, so if one don't want to see only the sidebar. The shortcut 1 is more for testing functionality change, like shortcuts on the webpage, while the shortcut 2 is more for testing content change. The starting point of shortcut 4 is editor/sidebar, else the shortcut 4 creates too much confusion.

Remark
This report is based on the knowledge of https://github.com/xiota/geany-preview/issues/37, https://github.com/xiota/geany-preview/issues/38 and https://github.com/xiota/geany-preview/issues/40.
The shortcut 2 exists already https://github.com/xiota/geanylua-scripts/tree/main/toggle-editor as stand-alone-version (,so without state file; the state file is mentioned in https://github.com/xiota/geany-preview/issues/37).

**Comment by xiota** [comment]
My understanding of what is possible now:

* Toggle editor visibility with a lua script.
* Toggle sidebar visibility with built-in geany shortcut.

The problem with the above is that both the editor and sidebar can be made not visible at the same time.  Also, I know how to change visibility in lua, but cannot change focus.  So with that in mind...  What about:

* If only editor is visible (no visible sidebar) and toggle editor key is pressed, then ... do nothing?  (Hide editor only when sidebar is visible.)
* If only sidebar is visible (no visible editor) and toggle sidebar key is pressed, then ... do nothing?  (Hide sidebar only when editor is visible.)
* New shortcut that changes views.  Editor only -> Editor + Sidebar -> Sidebar only -> Editor + Sidebar -> Editor only -> ...

**Comment by ralf3u** [comment]
>..., but cannot change focus.

This stands in contradiction to https://github.com/xiota/geany-preview/issues/5#issuecomment-1272538526.

**Comment by xiota** [comment]
You've opened up a lot of issues with long descriptions.  It's easy to misread or lose track of what's going on.  I don't know how to change focus in Lua, and if I had previously indicated otherwise, it was because I had forgotten or been mistaken about whether focus (or visibility) was the topic of discussion.  You can probably point back to more posts where this or that was said, and it would just make everything even more complicated and easier to misread and lose track of.

While changing focus among components is possible to an extent in C/C++, the exact component you want to have focused is often missed (eg, issues you've opened where various navigation steps don't do what's expected).  Arbitrary focus changes beyond what's already done (change focus between editor and preview) are outside the scope of the Preview plugin.  The plugin is intended to preview markup, not replace the web browser.

If you can provide a clear, *brief* description of a *few* shortcuts for changing focus, I can think about adding them to the xi-tweaks or tweaks-ui plugins.

**Comment by ralf3u** [comment]
>The first one is named "Toggle between only editor and editor/sidebar". What is toggled? Visibility? Focus? Something else?

Yes, you are right, the names are not clear. So, here the new names:
Shortcut names
1: `Toggle visibility/focus between editor-only and editor/sidebar`
2: `Toggle visibility between editor-only and editor/sidebar; editor stays in focus`
3: `Toggle visibility/focus between sidebar-only and editor/sidebar`
4: `Toggle visibility/focus between editor-only and sidebar-only`
5: `Exit "Toggle visibility/focus between editor-only and sidebar-only"`

to 5: I prefer exit than restore

>The description is long and confusing.

Sorry, I wrote the description in that way how I would program it in JavaScript.

So, I will try a new description for 1:
editor/preview is visible; I don't want to be disturbed by the preview while coding, so I press shortcut 1; now only the editor is visible and the editor is in focus, so I can code without preview; then I want to test the functionality of the webpage, so I press again the shortcut 1; now editor/preview is visible and preview is in focus

Is it OK like this, so I would continue to describe the shortcuts 2, 3, 4 and 5 in that style?

> I've proposed to combine the latter two into a single shortcut that switches among views: editor-only, editor+sidebar, sidebar-only.

Well, at the moment I'm not so interested in that shortcut.

>Toggle editor visibility with a lua script.
Toggle sidebar visibility with built-in geany shortcut.
The problem with the above is that both the editor and sidebar can be made not visible at the same time.

Yes, that was the reason to create an own shortcut for the sidebar and to add a state file, what is mentioned in the report, to check first if the other shortcut is active or not. If it is active, then I think it needs first to be disabled.

**Comment by xiota** [comment]
I've added a set of scripts, switch-views.  They don't do exactly what's described above, but you should try them before making them more complicated.  Also updated the toggle-editor and auto-sidebar scripts so that they all use the same functions as the switch-views scripts.

**Comment by ralf3u** [comment]
>1: Toggle visibility between editor-only and editor+sidebar
2: ...
3: Toggle visibility between sidebar-only and editor+sidebar
4: Toggle visibility between editor-only and sidebar-only
5: Show both editor and sidebar

Two things that I would change:
- 1 & 3: instead of `editor+sidebar`, I would use of `editor/sidebar` because it's compact; the slash reminds me of the line between the editor and the sidebar
- 5: instead of `Show both editor and sidebar`, I would prefer `Exit "Toggle visibility between editor-only and sidebar-only"`

>5 could use some explanation. You use the word "exit", but that would mean you have to "enter" some sort of mode to use 1-4.

Exit is only for leaving 4.  I think it's better to have a clear separation of the shortcuts, else it creates confusion . So, let's say I press the shortcut 4 multiple times for many hours. I can imagine that after that, if I want to see editor/sidebar again, I have difficulties to find a way how to get out of this loop. The word "Exit" would be a good help in my head for leaving this loop.

>I think a single shortcut to switch among all three views would be the most useful because it effectively provides the functionality of multiple shortcuts with reduced cognitive load.

Well, I think that a clear separation of shortcuts from shortcut 1 to shortcut 5 would be the most useful, even if the cognitive load will increase.

>They don't do exactly what's described above, but you should try them before making them more complicated.

I will try them tomorrow and I will give you feedback.

**Comment by xiota** [comment]
> I will try them tomorrow and I will give you feedback.

Thanks.  It will be easier to modify something concrete.  I just pushed an update.  The shortcuts now are:

* Toggle editor visibility
* Toggle sidebar visibility
* Show both editor and sidebar
* Switch views (editor → editor+sidebar → sidebar)
* Switch views (editor ↔ sidebar)

They can be renamed by making symlinks and editing `hotkeys.cfg`.

While coding the above, I just saw some Lua functions that *might* focusing the editor (but not the sidebar).  However, I want to finalize the visibility portion before trying to change focus.

> instead of `editor+sidebar`, I would use of `editor/sidebar`

The problem is the name of the Lua shortcuts are based on the file name.  Linux uses `/` as path separator, so it can't be used in the file name.

> I can imagine that after that, if I want to see editor/sidebar again, I have difficulties to find a way how to get out of this loop.

The way to "exit" the loop is to use a different shortcut.  Suppose you're using a shortcut that alternates views between editor-only and sidebar-only.  Now you stop at sidebar-only and want to show both editor and sidebar.  You can toggle editor visibility.  Then both will be visible.

**Comment by ralf3u** [comment]
Just for documentation: The shortcuts are located in

https://github.com/xiota/geanylua-scripts/tree/main/switch-views
and in
https://github.com/xiota/geanylua-scripts/tree/main/toggle-editor

Before testing: Is there a reason why toggle-editor is not part of switch-views?

>The way to "exit" the loop is to use a different shortcut. Suppose you're using a shortcut that alternates views between editor-only and sidebar-only. Now you stop at sidebar-only and want to show both editor and sidebar. You can toggle editor visibility. Then both will be visible.

Yes, that's right. That was also an option for me when developing the shortcuts in my head.

On the webpage https://github.com/xiota/geanylua-scripts/tree/main/switch-views, there is:
switch-views-(editor-→-editor+sidebar-→-sidebar).lua
I would not call it like this, because in the address bar of Firefox the + is not shown:
switch-views-(editor-→-editor%2Bsidebar-→-sidebar).lua

Instead of a plus, one could use a hyphen.

**Comment by xiota** [comment]
> Is there a reason why toggle-editor is not part of switch-views?

toggle-editor is included with switch-view.

> in the address bar of Firefox the + is not shown

This is for Geany/Lua.  Why does Firefox matter?

> Instead of a plus, one could use a hyphen.

Hyphens are converted to spaces in the shortcut name.

**Comment by ralf3u** [comment]
>toggle-editor is included with switch-view. There's no need to use the separate toggle-editor. That folder exists because it was written first.

The new toggle-editor is different than the toggle-editor that you offered some weeks ago. Can the old toggle-editor still be used? What is the difference between the old and the new toggle-editor?

>This is for Geany/Lua. Why does Firefox matter?

It could be that other users do it like me:
They let appear the code in RAW. Then they copy the code from the browser and paste the code in the editor. Then they save the document by giving a name. For the name they copy a part of the name of the URL to paste it in the save-dialog. Now if someone is not so careful, the person will not recognize that it is not the right name.
If you think that a hyphen is not good, an alternative for the +-symbol could be: `_and_`.

I just tried all shortcuts. All shortcuts don't work. The message for the first shortcut:

![geany_preview_error](https://user-images.githubusercontent.com/68117017/222090247-ab695a88-fc4d-4830-87e7-c42fcd9a0890.jpg)

For the other shortcuts it is nearly the same message. Only the names of the documents are different.

Just for information:
Then I deleted all files from geanylua in order to paste toggle-editor.
What works is toggle-editor from https://github.com/xiota/geanylua-scripts/tree/main/.

**Comment by xiota** [comment]
> They let appear the code in RAW. Then they copy the code from the browser and paste the code in the editor.
> ...
> I just tried all shortcuts. All shortcuts don't work.

The shortcuts use symlinks.  You can either recreate them with whatever name you want or install using the following steps:

* Download from here: https://github.com/xiota/geanylua-scripts/archive/refs/heads/main.tar.gz
* Copy the file to `/tmp` and unpack with `tar`:
   ```
   cp ~/Downloads/geanylua-scripts-main.tar.gz /tmp
   tar xz /tmp/main.tar.gz`.
   ```
* Copy the files into the Geany lua plugins folder: `~/.config/geany/plugins/geanylua/`

**Comment by ralf3u** [comment]
```
cp ~/Downloads/geanylua-scripts-main.tar.gz /tmp
tar xz /tmp/main.tar.gz`.
```
First command did work, but not the second command, even not with this command: 
`tar xz /tmp/geanylua-scripts-main.tar.gz`
I used right-click and `Extract Here`.

I copied the files into geanylua, but no shortcuts appear for Lua Scripts. In geanylua there are the folders auto-sidebar, column-markers and so on.

>The shortcuts use symlinks. You can either recreate them with whatever name you want or ...

I have no idea what symlinks are. Could you please give an example how to recreate them?

**Comment by xiota** [comment]
Sorry, wrote the command wrong.  Should be:
```
cd /tmp
tar xf geanylua-scripts-main.tar.gz
```

Copy the files from inside the `switch-views` folder into  `~/.config/geany/plugins/geanylua/`.

Symlinks are references to the real files.  This way the keybinding can be renamed without affecting the actual script.  They're created with the command:
```
ln -s [source] [link]
```

**Comment by ralf3u** [comment]
Oh, that's really really cool!!!!!

I like specially switch-views-(editor-↔-sidebar).lua. It's so smooth!! It feels better than I expected!
The shortcuts are a big release for me, because now I could use Geany Preview also on small displays.
I still don't see a use for me in the shortcut switch-views-(editor-→-editor+sidebar-→-sidebar).lua, but maybe somebody else will like it. I think that it is better to have more options than less options.

As far as the shortcuts concerned, I couldn't find bugs, so it's perfect.

Issues with the scroll-bar within the sidebar
I can see on three shortcuts the scroll-bar in the sidebar for a part of a second.
1
toggle-editor-visibility.lua
If editor and sidebar are both visible, and if I press the shortcut, then there is the scroll-bar in the sidebar for a part of a second. I can see it every second press. I can see it also when holding the press.
2
switch-views-(editor-↔-sidebar).lua
If editor and sidebar are both visible, and if I press the shortcut, then I can see only in the second press the scroll-bar in the sidebar for a part of a second. So that means that then if I hold the press, then the scroll-bar in the sidebar is not visible.
3
switch-views-(editor-→-editor+sidebar-→-sidebar).lua
If editor and sidebar are both visible, and if I press the shortcut, then I can see the scroll-bar in the sidebar for a part of a second. I can see it every third press. I can see it also when holding the press.

Focus
If the focus-part would work, then this would be just amazing.
Because the focus-part does not work, there are a lot of issues. Here are two of them:
1
If editor and sidebar are both visible, and if I press the shortcut switch-views-(editor-↔-sidebar).lua, then the tab-title is in focus, what means that, if I press the left-arrow-key, then the content of the Preview-tab will not be visible any more, but the content of the tab what is on the left side of the Preview-tab.
2
If editor and sidebar are both visible, and if I press 3x the shortcut switch-views-(editor-↔-sidebar).lua, and then, if I press the down-arrow-key on the keyboard, and then if I press the left-arrow-key on the keyboard, then in the editor the tab on the left side will be in focus.

Based on the number of documents and code, I can imagine that you worked a lot for the realization of the shortcuts. Thank you so much for the shortcuts.

**Comment by xiota** [comment]
> I can see on three shortcuts the scroll-bar in the sidebar for a part of a second.

This is out of my control.

> Focus

There is *potential* to focus the editor, but not the sidebar.

> If editor and sidebar are both visible, and if I press the shortcut switch-views-(editor-left_right_arrow-sidebar).lua, then the tab-title is in focus, what means that, if I press the left-arrow-key, then the content of the Preview-tab will not be visible any more, but the content of the tab what is on the left side of the Preview-tab.

Are you making the editor invisible?  Then the focus goes to the sidebar tabs?  As far as I can tell, this may not be fixable.

> If editor and sidebar are both visible, and if I press 3x the shortcut switch-views-(editor-left_right_arrow-sidebar).lua, and then, if I press the down-arrow-key on the keyboard, and then if I press the left-arrow-key on the keyboard, then in the editor the tab on the left side will be in focus.

This *might* be fixable by focusing the editor.  I plan to test this weekend.

> Thank you so much for the shortcuts.

You're welcome.

**Comment by xiota** [comment]
I just did some testing.

* Unable to replicate your focus issues.  When I press left/right/down, keyboard focus moves into the editing area.  The editor tabs do not change.
* The lua functions that I thought might change widget focus don't seem to work.  I'm going to open a bug report at geany/geany-plugins.

**Comment by ralf3u** [comment]
There's somewhere a bug. I deleted all files in geanylua. Now if a HTML-document is in focus and if I let appear the sidebar through the menu bar (View), then the editor is not there any more. Only the sidebar is visible. Normally editor and sidebar should be visible side by side. I think it has maybe to do something with this: I deleted the swich-views-files at the wrong moment (maybe Geany was open, but I don't think so), when only the sidebar was visible. What do you think?

**Comment by ralf3u** [comment]
I just pasted all switch-files in geanylua. All Geanylua shortcuts work, but not correctly. Maybe it has to do with my last comment and with the created file `toggle_sidebar-hidden`?

**Comment by ralf3u** [comment]
I can't see any more editor and sidebar side by side, even if the geanylua-file is empty, even after a restart of the computer, even after activating the sidebar.

**Comment by ralf3u** [comment]
>Are you using the lua toggle-sidebar and not the built-in toggle sidebar?

In the Preferences of Geany in Keybindings there is no shortcut for `Toggle Sidebar`.

>Are you copy/pasting from the webpage or copying the files from the tar.gz? Make sure you use the tar.gz, not the zip because zip may not support symlinks. If the problem is the symlinks, I may change it to not use symlinks.

I did this:
```
cp ~/Downloads/geanylua-scripts-main.tar.gz /tmp
cd /tmp
tar xf geanylua-scripts-main.tar.gz
```

>Maybe check events/init.lua ?
```
-- load and initialize toggle_editor and toggle_sidebar scripts
local lua_path = geany.appinfo().scriptdir..geany.dirsep

--[[
  options:
    restore - startup with the last editor state
    hide   - startup with editor hidden
    *show    - startup with editor visible
    toggle  - startup with the editor toggled from the last editor state
--]]

editor_start = "show"
sidebar_start = "show"

toggle_editor = loadfile(lua_path.."toggle-editor.lua")
toggle_editor()

toggle_sidebar = loadfile(lua_path.."toggle-sidebar.lua")
toggle_sidebar()
```

I deleted all files from the geanylua-file to test the old auto-sidebar. Also with the old auto-sidebar I can't see any more editor and sidebar side by side.

**Comment by xiota** [comment]
> In the Preferences of Geany in Keybindings there is no shortcut for Toggle Sidebar.

Preferences / Keybindings / Lua Script / Toggle sidebar visibility

> I did this...

Then to complete installation, open `/tmp/geanylua-scripts-main` in your file manager and copy the desired files into `~/.config/geany/plugins/geanylua`.

> I can't see any more editor and sidebar side by side

Editor state should be restored on Geany restart.

Toolbar needs to be toggled on from Geany perspective.  Menubar / View / Show Sidebar.

**Comment by ralf3u** [comment]
>Preferences / Keybindings / Lua Script / Toggle sidebar visibility

Yes, there is a shortcut. It is F4. If I toggle F4, then editor is shown or sidebar is shown, but not both at the same time.
If I press the shortcut "Show both editor and sidebar", then there is only the sidebar, no matter how often I press the shortcut.

>Then to complete installation, open /tmp/geanylua-scripts-main in your file manager and copy the desired files into ~/.config/geany/plugins/geanylua.

Yes, this is what I did.

> Menubar / View / Show Sidebar.

Yes, this is what I did.

**Comment by ralf3u** [comment]
Maybe this could help to find a solution:
If I deactivate the plugin Geany Preview in the Plugin Manager, and then, if I activate the `Show Sidebar` in View in the menu-bar, then the editor disappears.

**Comment by ralf3u** [comment]
What I did today: I tested the switch-views, then deleted the files, then I tested the old auto-sidebar, then I deleted the files, then I wanted to test again the switch-views by pasting the files. And then the bug was there.

Just for information: I will continue tomorrow. There is no hurry to fix this problem the next days.

**Comment by ralf3u** [comment]
I'm so sorry. The line (between editor and sidebar) was completely at the right, nearly invisible. I have no idea how this happened. So, everything is fine. Again: I'm so sorry. Thank you for your help.

**Comment by ralf3u** [comment]
>Unable to replicate your focus issues. When I press left/right/down, keyboard focus moves into the editing area. The editor tabs do not change.

I can  not reproduce 1. I don't know why. But I can reproduce 2 in that way:

If editor and sidebar are both visible at the same time, then I click in the middle of the content area of the editor, and if I press 3x the shortcut `Switch views (editor-sidebar)` then, if I press the down-arrow-key on the keyboard, and then if I press the left-arrow-key on the keyboard, then in the editor the tab on the left side will be in focus.

**Comment by xiota** [comment]
> The line (between editor and sidebar) was completely at the left, nearly invisible. I have no idea how this happened. So, everything is fine. Again: I'm so sorry. Thank you for your help.

I was thinking about suggesting you check for that.  Glad you found it.  I think Geany has some bugs related to sizing the panels when the window is resized/maximized.  I think there's already a bug report, but it's unlikely to be fixed any time soon because the interaction between the components is complicated and it's hard to change anything without breaking something else.

> But I can reproduce 2 in that way...

Unable to replicate on my computer.  In menbar / Help / About, what is the version of Geany and GTK/GLib?  I have Geany 1.38, GTK+ 3.24.36, GLib 2.74.6.

Maybe related to the desktop/window manager?  I am using KDE Plasma 5.27.2.  I will try with XFCE to see if there is any difference.

**Comment by ralf3u** [comment]
>I think Geany has some bugs related to sizing the panels when the window is resized/maximized.

Aaahh ...! Yes, resize! That explains why the line between editor and sidebar disappeared. I was wondering because I didn't change the width of the editor.
I could reproduce the issue:
only-editor-view, then window 50% (width that starts on the left side), close Geany, delete all switch-files in geanylua, reopen Geany, maximize Geany, close Geany, paste again the switch-files in geanylua, reopen Geany

>I think there's already a bug report, ...

For documentation reasons, do you have the issue number of the bug?

>>If editor and sidebar are both visible at the same time, then I click in the middle of the content area of the editor, and if I press 3x the shortcut Switch views (editor-sidebar) then, if I press the down-arrow-key on the keyboard, and then if I press the left-arrow-key on the keyboard, then in the editor the tab on the left side will be in focus.
>>
>The problem seems to be related to message window visibility. I have the message window visible. When I hide it, I am able to replicate your issue. Since I cannot change focus with Lua, a workaround would be to show the message window (menu-bar / View / Show Message Window) and resize it so that it's hidden.

Well, at the moment this issue is not so important for me. I handle the problem like this: after switching: I press 1x the arrow-right-key, so that the cursor is visible in the editor.


To the focus problem 1 mentioned in https://github.com/xiota/geanylua-scripts/issues/7#issuecomment-1451950479:
>If editor and sidebar are both visible, and if I press the shortcut switch-views-(editor-left_right_arrow-sidebar).lua, then the tab-title is in focus, what means that, if I press the left-arrow-key, then the content of the Preview-tab will not be visible any more, but the content of the tab what is on the left side of the Preview-tab.

Sorry, I recognized that I did some mistakes, so here is the correction:
If editor and sidebar are both visible, and if I click in the middle of the editor, and if I press 1x the shortcut `Toggle editor visibility`, then if I press 1x first the bottom-arrow-key and then 1x the left-arrow-key, then the content of the Preview-tab will not be visible any more, but the content of the tab what is on the left side of the Preview-tab.
To avoid the issue, I press 2x the bottom-arrow-key, so that the content of the Preview-tab is in focus.
So, in total I handle the focus problem like this:
For only-editor: If the cursor is not visible, I press 1x the arrow-right-key.
For only-sidebar: If I want to put the content of the Preview-tab in focus I press 2x the bottom-arrow-key.

>I can see ... the scroll-bar in the sidebar for a part of a second.
>>
>This is out of my control.

The scroll-bar that is visible for a part of the second is more on the left side, although my scroll-bar of the editor is more on the right side when editor and sidebar are both visible (I use the editor on the left side, while the sidebar is on the right side). That's strange. From the color of the scroll-bar and the length of the inside of the scroll-bar it looks like it is the same scroll-bar like the scroll-bar of the sidebar that is located completely on the right side of the window. That's also strange. Is that OK if I open a new issue at Geany, or you want to open it?

Just for documentation: I tested the shortcuts with an HTML-document and with a Fountain-document. 

Before closing this issue I would like to test the shortcuts with the new auto-sidebar what I will do in the next days. I will leave a feedback here.

**Comment by xiota** [comment]
> >  I think Geany has some bugs related to sizing the panels when the window is resized/maximized. I think there's already a bug report...
>
> For documentation reasons, do you have the issue number of the bug?

No, but searching Geany issues for "[resize](https://github.com/geany/geany/issues?q=is%3Aissue+is%3Aopen+resize)" shows multiple related results.

> ... here is the correction...

Okay.  I am able to replicate that.  Unfortunately, focus issues are out of my control with Lua.

> ... scroll-bar that is visible for a part of the second...

The problem is probably related to the timing of screen updates and hiding widgets along with the speed of your computer.  It's just a graphical glitch, and I don't think there is anything that can be done about it.  This script hides things in ways that Geany wasn't necessarily designed to handle.  This should *not* be reported to Geany unless it can be replicated using only built-in Geany functions.

> I would like to test the shortcuts with the new auto-sidebar... I will leave a feedback here.

Will look forward to it.

**Comment by ralf3u** [comment]
>While coding the above, I just saw some Lua functions that might focus the editor (but not the sidebar). 

>The lua functions that I thought might change widget focus don't seem to work. I'm going to open a bug report at geany/geany-plugins.

For reasons of documentation I mention the report: https://github.com/geany/geany-plugins/issues/1229.
Is the report just about the focus in the editor or is the report also about the focus in the sidebar?

**Comment by xiota** [comment]
@ralf3u Please don't *link* Geany issues here or vice versa because some Geany devs have expressed irritation in the past.  It's fine if you can refer to them without GitHub creating the link.

**Comment by ralf3u** [comment]
>The lua functions that I thought might change widget focus don't seem to work. I'm going to open a bug report at geany/geany-plugins.

>Please don't link Geany issues here or vice versa because some Geany devs have expressed irritation in the past. It's fine if you can refer to them without GitHub creating the link.

OK. So, for reasons of documentation I mention the issue number of the bug report at geany/geany-plugins: 1229.

>If you read the comments at the issue you had linked, you'll see that elextr had stated, "Even if it did set the notebook tab [GTK](https://docs.gtk.org/gtk3/method.Notebook.set_current_page.html) does not say it grabs focus, so the Geanylua doc... is wrong anyway."

Sorry, but I don't understand anything of what is written in the whole issue. So, is the report just for the focus in the editor or is the report also for the focus in the sidebar? And is there a conclusion, like even if the GeanyLua would work correctly, it could have no effect on the focus?

**Comment by xiota** [comment]
The report is that `geany.activate` doesn't work.  `geany.activate` calls some C code that switches tab.  The documentation for `geany.activate` says it also changes focus.  But GTK documentation of the underlying C function doesn't mention focus.  So focus is a *possible*, but not guaranteed, side effect.

I've fixed the issue and opened a PR geany/geany-plugins/1234.  I haven't tested, but suspect focus will not do what you want.

**Comment by ralf3u** [comment]
>I've fixed the issue and opened a PR geany/geany-plugins/1234. I haven't tested, but suspect focus will not do what you want.

Thank you so much for your work and for the attempt to  fix the focus problem.

**Comment by ralf3u** [comment]
Correct me if I'm wrong. In the preferences of Geany, in the Action `Focus`, there are two shortcuts:
`Switch to Editor`, and
`Swich to Sidebar`.
They work absolutely perfect. In both shortcuts the content-area of the tab is in focus. So, what about to copy and paste them formula in Geanylua for all shortcuts?

**Comment by xiota** [comment]
Thanks for pointing those out.  I just noticed Lua has access to those functions.  The commands are (writing here for future reference):
```
geany.keycmd("FOCUS_EDITOR")
geany.keycmd("FOCUS_SIDEBAR")
```
I'll look at them more closely later.

**Comment by ralf3u** [comment]
>I'll look at them more closely later.

Would it be possible to solve at the same time #8?

**Comment by ralf3u** [comment]
My suggestions for the names and the orders of the shortcuts:

I would not mention `visibility` and I would not mention `focus`, except in one case, else it makes everything complicated.
I would not write the exit-shortcut in the middle of the list, but at the end of the list, because it is a kind of Reset for all shortcuts.

My perfect list would be:

Switch editor ↔ editor+sidebar
Switch editor ↔ editor+sidebar; focus stays in the editor
Switch sidebar ↔ editor/sidebar
Switch views (editor → editor+sidebar → sidebar)
Switch views (editor ↔ sidebar)
Show editor+sidebar

**Comment by ralf3u** [comment]
>Are you still using an Ubuntu variant? Which version? 

Lubuntu 22.04.1, on an other computer it is Lubuntu 22.10.

> Since you've mentioned that your computer is slow ...

One computer is now fast (intel i3 11th generation 16GB RAM; Intel UHD Graphics; Lubuntu 22.04.1). I saw the sidebar glitch on the fast computer.

>The plugins are on the [PPA](https://launchpad.net/~xiota/+archive/ubuntu/geany-plugins), available for 22.04, 22.10, 23.04.

I just do `sudo apt update` and then `sudo apt dist-upgrade`. Is that correct?

**Comment by ralf3u** [comment]
>I just do sudo apt update and then sudo apt dist-upgrade.

Then I restarted Geany. The sidebar glitch is still there. No difference.
On the slow computer I could install an other system for testing the sidebar glitch on an other system, if you want. Just tell me which kind of system.

**Comment by xiota** [comment]
> The sidebar glitch is still there. No difference.

I didn't expect there would be any difference.  Just wanted to try.

Don't worry about the graphical glitch anymore.  I think there's nothing reasonable that can be done about it.  Would probably require refactoring the GTK toolkit so that components can all be hidden/shown and blitted to the screen within a single frame.

> On the slow computer...

Do Lua scripts seem any faster in general on the slow computer?  Or no difference?

**Comment by ralf3u** [comment]
I would not call it sidebar-glitch, but the-scrollbar-of-the-sidebar-glitch.
On the slow computer it is also there, without difference.

Today I looked closely at the shortcut `Switch editor ↔ editor+sidebar`. Then I understood what is happening:
Let's assume that the sidebar is on the right side, with a width of 30%. In the moment where I press the button, the sidebar is moving completely to the left side of the Geany-window, what means that the scrollbar of the sidebar is not any more completely on the right side of the Geany-window, but 30% from the left side of the Geany-window. That is also happening not only with the scrollbar of the sidebar, but also with the content of the tab of the sidebar, so for example also the text is moving from the right side to the left side. Then the scrollbar disappears and then the content of the tab will be stretched to the right border of the Geany window. That is also why sometimes I can not only see the scrollbar for a part of a second but also text at the left bottom for a part of a second before the text will be stretched to the right side.

I made some tests with the corresponding shortcut in the Geany-built-shortcut `Toggle sidebar`. There I can't see the-scrollbar-of-the-sidebar-glitch, even not when I hold the shortcut. The question is: Why does the issue exists in the Geanylua-shortcut but not in the Geany-built-shortcut, although both should do exactly the same?

Could you please tell me what exactly is happening, when the Geanylua-shortcut is pressed in slow motion, so step by step, so I can understand every single step? Maybe I could have some ideas to solve the issue.

If in the sidebar there is no scrollbar, because the text is short, then the issue does not exist.
Till now we are talking about the computers I use, but do you see also the-scrollbar-of-the-sidebar-glitch on your computer?

**Comment by xiota** [comment]
Are you able to make a video and temporarily post it on youtube or somewhere, so I can see what is happening?

**Comment by ralf3u** [comment]
Sorry, in my last comment I noticed the wrong shortcut. It is not the shortcut `Switch editor ↔ editor+sidebar`. It is the shortcut `Switch sidebar ↔ editor+sidebar`. I just did the correction in the comment.

>Are you able to make a video and temporarily post it on youtube or somewhere, so I can see what is happening?

Today I will try to find a way.

**Comment by ralf3u** [comment]
>Are you able to make a video and temporarily post it on youtube or somewhere, so I can see what is happening?

rudolph.bplaced.net
There you find the file geanylua.mov. Right-click on it and then select "Save Link As...".
Once it is completely downloaded, look in the file-manager where it is on your computer, and then let it play in the VLC-player.

**Comment by ralf3u** [comment]
I just made a test with the old toggle-editor version from last year. The scrollbar-of-the-sidebar-glitch is also there. It was mentioned for the first time in https://github.com/xiota/geanylua-scripts/issues/4#issuecomment-1152776068 in point 3.

**Comment by xiota** [comment]
From the video, looks similar to what I imagined.  Doesn't happen on my computer.  Are any other plugins active besides Preview and GeanyLua/switch-views?  Do you see the same thing with the built-in toggle sidebar?

**Comment by ralf3u** [comment]
>Are any other plugins active besides Preview and GeanyLua/switch-views?

No.

>Do you see the same thing with the built-in toggle sidebar?

No.

**Comment by xiota** [comment]
I just updated switch views.  Files are reorganized, so please clear out the plugins/geanylua folder.  Then [redownload](https://github.com/xiota/geanylua-scripts/archive/refs/heads/main.zip) and unzip the new files.

- The shortcuts change focus between the editor and sidebar.
- The symlinks are replaced with regular files to make them easier to manage.
- Changed how often the `update()` function is called.  Might affect the scrollbar glitch.  (Doubtful through.)
- You can rename the shortcuts later by renaming the files and editing `hotkeys.cfg`.

**Comment by ralf3u** [comment]
I'm sorry. I just recognized that I did an other mistake in the [comment](https://github.com/xiota/geanylua-scripts/issues/7#issuecomment-1458941359) of yesterday. The shortcut `Toggle sidebar` in Geany does something else than `Switch sidebar ↔ editor+sidebar`. There does not exist a corresponding shortcut in Geany.
It was late when I was reporting. I just crossed out some text in https://github.com/xiota/geanylua-scripts/issues/7#issuecomment-1458941359 and in https://github.com/xiota/geanylua-scripts/issues/7#issuecomment-1460199799.

**Comment by ralf3u** [comment]
I unziped the files and pasted them in geanylua. When I open Geany the answer is:
![answer2](https://user-images.githubusercontent.com/68117017/223791680-8d9e8686-fc8e-4f22-86b9-d97c7a082f34.jpg)

**Comment by xiota** [comment]
Forgot to update `switch-views/update events/init.lua`.  It should be fixed now.

**Comment by ralf3u** [comment]
Show both editor and sidebar:
![geanylua](https://user-images.githubusercontent.com/68117017/223835787-0accc66f-91c9-4a95-9be8-f2dfdb151dfc.jpg)

**Comment by ralf3u** [comment]
shortcut test does not work
Switch view (editor/sidebar): sidebar is not in focus
Toggle sidebar 2: sidebar is not in focus

**Comment by xiota** [comment]
> Show both editor and sidebar

Fixed.

> shortcut test does not work

That's a test shortcut.  Not supposed to be there.  Removed.

> ... sidebar is not in focus

Added two new shortcuts for testing: focus-editor and focus-sidebar.

**Comment by ralf3u** [comment]
glitch still there

Focus editor:
![focus_editor](https://user-images.githubusercontent.com/68117017/223855938-76728c24-f292-495e-b2da-81df702d41c4.jpg)

Focus sidebar:
![focus_sidebar](https://user-images.githubusercontent.com/68117017/223855991-4336919d-c5f0-4339-b334-18fd8f798d6b.jpg)

**Comment by ralf3u** [comment]
Toggle sidebar 2: focus does not work for the sidebar
Switch view (editor/sidebar): sidebar is still not in focus

**Comment by xiota** [comment]
> Focus editor... Focus sidebar...

Should be fixed.

> Toggle sidebar 2... Switch view (editorleft_right_arrowsidebar)... sidebar is still not in focus

That is what the focus-editor and focus-sidebar scripts are testing, whether just changing focus by itself works.  Possible that showing/hiding the sidebar then trying to focus it won't work.  Maybe not enough time between changing visibility and trying to focus?  Or something else in Geany is interfering.

**Comment by ralf3u** [comment]
Shortcut `Toggle sidebar 2`
After having pressed 1x the shortcut, and if I press 1x the left-arrow-key, then the focus is in the editor, although it should still be in the content-area of the Preview-tab.

Shortcut `Toggle sidebar 2` OR shortcut Switch view (editor↔sidebar)
a
After having pressed 1x the shortcut, the sidebar is in focus, but not the content-area of the Preview-tab. I can recognize that by tab-browsing with Ctrl+Page-up/Page-down in the sidebar. But I don't think that before tab-browsing that the tab-title is in focus, because if the tab-title would be in focus, then 1x press of bottom-arrow-key would put the content-area of the tab in focus (see Remark down below). What I think is: After having pressed 1x the shortcut, even if the sidebar is in focus, neither the content-area of the tab, nor the tab-title is in focus.
b
After having pressed 1x the shortcut, I have to press a special combination of the arrow-keys (bottom & top) till the content-area of the Preview-tab is in focus:
bottom, top, bottom
or
top, bottom
What I think is: The top-arrow-key is firing something. I think it fires that the tab-title is in focus, because if I press then 1x the bottom-arrow-key, the content-area of the tab-title is in focus.

Remark
The issue reminds me to https://github.com/xiota/geany-preview/issues/15. There after having clicked on the Preview-tab-title, I need to press one time the bottom-arrow-key, so that the content-area of the Preview-tab is in focus.

Shortcut `Focus sidebar`
It does not work with the shortcut `Switch view (editor↔sidebar)` and also not with the shortcut `Toggle sidebar 2`.

Possible solutions
a
If I press the Geany-Preview-shortcut `Toggle focus between editor and sidebar`, then it sets the focus correctly. What about to copy and paste the formula in Geanylua?
b
Shortcut `Toggle sidebar 2`: I can imagine that this could be solved like this: At the end of showing again editor+sidebar, add the command:
Put focus in the content-area of the sidebar.
c
Shortcut Switch view (editor↔sidebar): I can imagine that this could be solved like this: 
If sidebar is visible then fire 2x the Geany-Preview-shortcut `Toggle focus between editor and sidebar` (it works!).
d
Shortcut `Toggle sidebar 2`: I can imagine that this could be solved like this: 
At the end of showing again editor+sidebar, then fire 2x the Geany-Preview-shortcut `Toggle focus between editor and sidebar` (it works!).

**Comment by ralf3u** [comment]
>c
Shortcut Switch view (editorleft_right_arrowsidebar): I can imagine that this could be solved like this:
If sidebar is visible then fire 2x the Geany-Preview-shortcut `Toggle focus between editor and sidebar` (it works!).
d
Shortcut Toggle sidebar 2: I can imagine that this could be solved like this:
At the end of showing again editor+sidebar, then fire 2x the Geany-Preview-shortcut `Toggle focus between editor and sidebar` (it works!).

As comparison: If I press 2x the shortcut `Focus sidebar`, it does not work.

**Comment by xiota** [comment]
I will add another test script.  This one tests another focus method.  But the specific behavior with the sidebar is probably not possible with Lua.

> Shortcut `Focus sidebar` ... does not work... If I press 2x the shortcut Focus sidebar, it does not work.

Then the sidebar cannot be focused the way you want from Lua.  The problem is probably the focus command focuses the sidebar, but what you really want to focus is the sidebar child (the content, the preview).

> the Geany-Preview-shortcut `Toggle focus between editor and sidebar`... What about to copy and paste the formula in Geanylua?

Cannot be done in Lua.  Would require changes to GeanyLua itself.  I am not willing to make the change.  It is already difficult to get real bug fixes accepted into the plugin.  (I have pending pull requests to fix `geany.activate` and `geany.keygrab`.)

> Put focus in the content-area of the sidebar.

There is no such command in GeanyLua.  The commands to focus the sidebar do not search for the child (content-area).

> fire 2x the Geany-Preview-shortcut `Toggle focus between editor and sidebar`

I don't think plugins can access functions from other plugins.

**Comment by xiota** [comment]
>  `Unable to locate geany-git`

I had built `geany-git` only for 22.04 at that time for testing.  Do you want `geany-git` for 22.10?  Do you want the same git version or a new one (whatever is current)?

The plugins on the PPA are named `geany-plugins`.  They work with the standard `geany` package in the Ubuntu repositories.

> What about the idea to execute focus-sidebar-2.lua and then to execute 1x bottom-arrow-key...

I don't see any way to send keypresses with Lua.

> ... put only the two shortcuts in Geany Preview...

This is not a suitable shortcut for Preview.  It could be considered for one of the tweaks plugins, but I wouldn't write it any time soon.

**Comment by ralf3u** [comment]
>>... put only the two shortcuts in Geany Preview...
>>
>This is not a suitable shortcut for Preview.

`Switch editor ↔ sidebar` is a suitable shortcut for Preview because users with small displays will use permanently the shortcut (I'm a user with a small display).
As for the shortcut `Toggle sidebar 2`, this shortcut is indeed not important at all (that was the reason why I didn't suggest that shortcut in the list of all shortcuts).

>It could be considered for one of the tweaks plugins, but I wouldn't write it any time soon.

OK. I will wait. Maybe next year?

Tab-switcher in Geany
This does not belong to this issue. Yesterday in Geany I pressed the wrong shortcut and suddenly a tab-switcher appears, some kind of poup-up where all open tabs were listed vertically. It is Shift+F10. Did you know the shortcut?

>I had built geany-git only for 22.04 at that time for testing. Do you want geany-git for 22.10? Do you want the same git version or a new one (whatever is current)?

The same version would be great (so with the scintilla patch); if it is too complicated then the current version would be nice, as long as the scintilla patch is part of it.

**Comment by ralf3u** [comment]
>I had built geany-git only for 22.04 at that time for testing. Do you want geany-git for 22.10? Do you want the same git version or a new one (whatever is current)?

Sorry, but at the moment I don't need a geany-git for 22.10 after all.
I changed from Lubuntu 22.04.1 to Lubuntu 22.10 specially because of a visual bug:
github.com/lxqt/lxqt/issues/2390#issuecomment-1462834203
In the task-switcher I can't see which task is in focus. It is really a big issue for me. In Lubuntu 22.10 it was not there, but today it is also there, so it can be that I will switch back to Lubuntu 22.04.1. I will make some further tests.

**Comment by xiota** [comment]
Okay.  Let me know if you want `geany-git` updated.  The latest git revision still has that bug where it doesn't switch to the desired document on startup.

**Comment by ralf3u** [comment]
>The latest git revision still has that bug where it doesn't switch to the desired document on startup.

The most important thing for me is the scintilla patch. Do you know if it is integrated in the current version? If yes, then I would know that the next version of Geany I can download Geany normally.

> Let me know if you want geany-git updated.

I could handle my issues in Lubuntu 22.10. But the problem is that it is supported only till July. So, then I will need also at the end of next month a geany-git for 23.04. But that version will be supported only till December. So, then I will need also the geany-git for Lubuntu 23.10 at the end of October. So, that are three different versions in one year. Do you say, `hey, that's too much`, what I can understand, then I will switch back to Lubuntu 22.04.1.

**Comment by ralf3u** [comment]
> I've updated the PPA for now, ...

Thank you so much for the geany-git.

This is what I did
```
sudo apt purge geany
```
Then I did:

```
sudo add-apt-repository ppa:xiota/geany-plugins
sudo apt update
sudo apt install geany-git

```
Then I did:

```
sudo add-apt-repository ppa:xiota/geany-plugins
sudo apt-get update
sudo apt-get install geany-plugin-preview

```
Then I opened Geany and typed some text. When closing Geany, Geany crashed.
I can't remember that Geany crashed the last days.
Do you want that I reinstall Lubuntu 22.10 to only install geany-git (so, without Geany Preview) and to test if that happens again? Or should I do some others tests?

>The most important thing for me is the scintilla patch.

I worked only some days with Lubuntu 22.10 and with Geany 1.38, but now I recognized that there was no scintilla issue at all. Maybe it is because I'm using an other computer than before. In the next days I would like to reinstall Lubuntu 22.10 and do the scintilla-test, because then it could be that I don't need any more a geany-git in the future.

**Comment by xiota** [comment]
> Geany crashed... Do you want that I reinstall Lubuntu 22.10 to only install geany-git... Or should I do some others tests?

No.  I think it may be related to upstream geany/geany issue 3095.  From comments, it looks like they're working on it.

> I worked only some days with Lubuntu 22.10 and with Geany 1.38, but now I recognized that there was no scintilla issue at all.

It's possible the packager included the patch to fix that specific bug.

> ... then it could be that I don't need any more a geany-git in the future.

That would probably be for the best.  `-git` packages can be pretty buggy.

**Comment by ralf3u** [comment]
Just for documentation to the two shortcuts that don't work properly:
>Maybe not enough time between changing visibility and trying to focus?

I don't think so, because in toggle-sidebar-2.lua I replaced the code
```
editor_focus()
sidebar_focus()

```

By this code:
```
if sidebar_is_visible() then
  sidebar_focus()
else
  editor_focus()
end

update()
```
If there would be not enough time, like you mentioned, then the focus would be always in the editor. But the focus is not always in the editor. So, for me, it is not a time problem. Which problem it is? I still don't know, but it is not the time.

I tried also this code:
```
if editor_is_visible() and sidebar_is_visible() then
  sidebar_focus()
end

update()
```
Also this does not work.

And also this does not work:

```
-- editor <-> sidebar

if editor_is_visible() then
  editor_hide()
else
  editor_show()
  sidebar_hide()
end

update()
```

And also this does not work:
```
-- editor <-> sidebar

if (editor_is_visible() then
  editor_hide()
elseif editor_is_visible() then
  editor_hide()
  sidebar_hide()
  editor_show()
  sidebar_show()
  editor_focus()
  if editor_is_visible() and sidebar_is_visible() then
  sidebar_focus()
  editor_hide()
  sidebar_focus()
  end
elseif sidebar_is_visible() then
  sidebar_hide()
  editor_show()
end

update()
```
What is interesting:
Although I created in the second IF exactly the same circonstances like the first IF, the focus in the sidebar is different. That is really strange.

**Comment by ralf3u** [comment]
>I don't think plugins can access functions from other plugins.

I just opened this question on geany geany discussions 3431 to be sure that it is not possible:
`Can plugins have access to functions of other plugins?`

**Comment by xiota** [comment]
> ... code ...

If you want to try to find a solution:

* `toggle-functions.lua` – Where those functions are implemented.
* [GeanyLua documentation](https://plugins.geany.org/geanylua/geanylua-ref.html) – Functions to control Geany.
* [Lua 5.1 short reference](https://plugins.geany.org/geanylua/luarefv51.html) – Lua language.

> I just opened this question on geany geany discussions 3431 ...  Can plugins have access to functions of other plugins?

I'll check it out in a few days to see what their response is.

**Comment by ralf3u** [comment]
>... 23.04 ... 23.10 ... 24.04 ...

I tested the scintilla-problem on an other computer with Lubuntu 22.10 and with Geany 1.38. Unfortunately the scintilla-problem still exists.

**Comment by ralf3u** [comment]
Just for information:
To the shortcut `switch views (editor vs sidebar)` I recognized that when only-sidebar is visible, if I tab-browse with shortcuts, then the content-area of the tab is in focus.
To the shortcut `switch views (editor vs sidebar)` I recognized that when only-sidebar is visible, if I press many times the top-arrow-key on the keyboard, then I navigate through the tool-bar.

To the shortcut `Toggle sidebar 2`
This behavior could be interesting: If I delete in toggle-sidebar-2.lua this line
```
editor_focus()
```
and save the document, then if I press the shortcut `Show both editor and sidebar`, and then, if I click in the content-area of the sidebar, and then, if I press 1x the shortcut `Toggle sidebar 2`, then if I press1x the bottom-arrow-key and then the right-arrow-key, I can navigate through the toolbar with the right-arrow-key.

**Comment by ralf3u** [comment]
>... 23.04 ... 23.10 ... 24.04 ...

I wanted to do myself a fix of scintilla (geany geany issues 3271 issuecomment-1236121036), so I'm independent of the geany-git, but I can't remember where the document editor.c is located to make the code change. Could you please tell me where the document editor.c is located?

In the comment there is this code, but I can't remember the document location:
```
/*#ifdef GDK_WINDOWING_QUARTZ*/
/*# if ! GTK_CHECK_VERSION(3,16,0)*/
	/* "retina" (HiDPI) display support on OS X - requires disabling buffered draw
	 * on older GTK versions */
	SSM(sci, SCI_SETBUFFEREDDRAW, 0, 0);
/*# endif*/
/*#endif*/
```

**Comment by ralf3u** [comment]
>the two shortcuts ... could be considered for one of the tweaks plugins, ...

I don't think that it is a good idea, because it makes everything complicated. The current shortcuts in geanylua should stay together because they work together and because there is also the auto-sidebar to respect.

>geany geany discussions 3431 discussioncomment-5356332

What about the idea to fork Geanylua, so to create some kind of GeanyLua special, that allows to have access to a common library?
So, then in Geany Preview the shortcut `Toggle focus between editor and the sidebar` needs to be removed and needs to be pasted in the library, so both plugins (Geanylua special and Geany Preview) have access to it.

What do you think about it?

**Comment by xiota** [comment]
> What about the idea to fork Geanylua...

Too much work to maintain a fork just for this.  Ideally the existing focus function in GeanyLua would be fixed, but I have other PRs with bug fixes pending that I would want to wait for before doing anything else with it.

**Comment by xiota** [comment]
> I wanted to do myself a fix of scintilla...

Are you referring to this: https://sourceforge.net/p/scintilla/bugs/2349/?page=1#f567 ?

The file is `geany/scintilla/gtk/ScintillaGTK.cxx`.  If you know a code snippet, you can find files with `grep`:
```
grep -R IS_WIDGET_MAPPED
```

**Comment by ralf3u** [comment]
To the auto-sidebar:

Last week this is what I did in Lubuntu 20.04.1:
In the file geanylua I pasted `activated.lua` in the file `events`.
In the existing document `init.lua` I pasted those lines:
```
auto_sidebar = loadfile(lua_path.."auto-sidebar.lua")
auto_sidebar()
```
I copy & pasted `toggle-sidebar.lua`.
I didn't pasted `toggle-functions.lua` because it is the same document.
In the document `hotkeys.cfg` I pasted this line:
`toggle-sidebar-visibility.lua`

When I wanted to start Geany, that was the answer:

![geanylua_autos-sidebar](https://user-images.githubusercontent.com/68117017/227469674-1a97a408-d172-439c-9817-f32e9d400979.jpg)

**Comment by ralf3u** [comment]
> scintilla fix

Did I understood it right?

I should not take this code:
```
/*#ifdef GDK_WINDOWING_QUARTZ*/
/*# if ! GTK_CHECK_VERSION(3,16,0)*/
	/* "retina" (HiDPI) display support on OS X - requires disabling buffered draw
	 * on older GTK versions */
	SSM(sci, SCI_SETBUFFEREDDRAW, 0, 0);
/*# endif*/
/*#endif*/
```

I should take this code, according to https://sourceforge.net/p/scintilla/bugs/2349/?page=1#f567:
```
   }
    if (IS_WIDGET_MAPPED(PWidget(wMain))) {
        ChangeSize();
+   } else {
+       const PRectangle rcTextArea = GetTextRectangle();
+       if (wrapWidth != rcTextArea.Width()) {
+           wrapWidth = rcTextArea.Width();
+           NeedWrapping();
+       }
    }

    alloc.x = 0;
```

Is this code replacing an other code or do I have only to paste the code somewhere.
If it replaces some code, could you please tell me which code will be replaced?
If I have only to paste the code, then where? Could you please add the code lines before and after the snippet above, so I know where it is located?

Thank you in advance.

**Comment by xiota** [comment]
> To the auto-sidebar...

I'll look at the [error](https://user-images.githubusercontent.com/68117017/227469674-1a97a408-d172-439c-9817-f32e9d400979.jpg) later.  Maybe sometime next week.  Busy this weekend.

> Did I understood it right?
>
> I should not take this code...
> ````
> /*#ifdef GDK_WINDOWING_QUARTZ*/
> /*# if ! GTK_CHECK_VERSION(3,16,0)*/
> 	/* "retina" (HiDPI) display support on OS X - requires disabling buffered draw
> 	 * on older GTK versions */
> 	SSM(sci, SCI_SETBUFFEREDDRAW, 0, 0);
> /*# endif*/
> /*#endif*/
> ````

I found that code verbatim in another file.  So it looks like the change has already been applied.

> I should take this code, according to https://sourceforge.net/p/scintilla/bugs/2349/?page=1#f567:
> ```
>    }
>     if (IS_WIDGET_MAPPED(PWidget(wMain))) {
>         ChangeSize();
> +   } else {
> +       const PRectangle rcTextArea = GetTextRectangle();
> +       if (wrapWidth != rcTextArea.Width()) {
> +           wrapWidth = rcTextArea.Width();
> +           NeedWrapping();
> +       }
>     }
> 
>     alloc.x = 0;
> ```

I think so.  It's in diff format.  The lines that start with a space are exactly in the file.  Lines starting with a `-` are in the original and will be removed.  Lines starting with a `+` will be added.  So look inside the file for code that looks like:
```
   }
   if (IS_WIDGET_MAPPED(PWidget(wMain))) {
       ChangeSize();
   }

   alloc.x = 0;
```

Then add the lines so it looks like:
```
   }
   if (IS_WIDGET_MAPPED(PWidget(wMain))) {
       ChangeSize();
   } else {
       const PRectangle rcTextArea = GetTextRectangle();
       if (wrapWidth != rcTextArea.Width()) {
           wrapWidth = rcTextArea.Width();
           NeedWrapping();
       }
   }

   alloc.x = 0;
```
Normally the changes will be in a `.diff` or `.patch` file that has the file names and line numbers so you can apply it with a `patch` command.  But without the extra info, you have to do it manually.

**Comment by ralf3u** [comment]
Sorry for my last comment. I will try first to install the nightly tarball of geany, and then I will see if the scientilla-patch was introduced.
I will leave you a message if I can't find it.

**Comment by ralf3u** [comment]
>I'll look at the error later. Maybe sometime next week. Busy this weekend.

Like everything, there's no hurry. I'm offline for about two weeks anyway.

**Comment by ralf3u** [comment]
> ... 23.04 ... 23.10 ... 24.04 ...

I just installed the nightly tarball of Geany in Lubuntu 23.04 by using my "step-by-step-description" at Geany issue-number 3271.

I have questions: In that step-by-step-description the last is command is:
`./geany -c ../config`
Is this necessary?
What does this command exactly mean?

Thank you for answers in advance.

**Comment by xiota** [comment]
> `./geany -c ../config`

* `./geany` – Run geany from the file located in the current directory.
* ` -c ../config` – Use custom config file located at `../config`;  The purpose is to test with a new config without modifyingan existing config.

Run `geany --help` to see more options.

**Comment by ralf3u** [comment]
If I open Geany from the home-file, then it is 1.39. If I open it from /usr/bin/geany then it is 1.38. Yesterday I just installed the nightly tarball of Geany and nothing else.
So, is this normal that I have two different versions although I downloaded only one file?

I tested the scintilla-bug in 1.38 that I downloaded yesterday and the issue is not there any more although the issue was in 1.38.
How can it be? Do you have an explication?

**Comment by xiota** [comment]
> If I open Geany from the home-file, then it is 1.39. If I open it from /usr/bin/geany then it is 1.38.

You must have two different versions in different locations.  The package manager usually installs to `/usr/bin`.  The other one depends on how you installed.  Could be `/usr/local/bin/`.  Might be somewhere else though.

**Comment by ralf3u** [comment]
On Lubuntu 23.04 I tried to install Preview on the nightly tarball of Geany with those lines:
```
sudo add-apt-repository ppa:xiota/geany-plugins
sudo apt-get update
sudo apt-get install geany-plugin-preview
```
like it is described on https://github.com/xiota/geany-preview.

The result in the terminal for the last line is:
```
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 geany-plugin-preview : Depends: libcmark-gfm0 but it is not installable or
                                 libcmark-gfm0.29.0.gfm.3 but it is not installable
                        Depends: libcmark-gfm-extensions0 but it is not installable or
                                 libcmark-gfm-extensions0.29.0.gfm.3 but it is not installable
E: Unable to correct problems, you have held broken packages.

```
Did I make something wrong?

**Comment by xiota** [comment]
According to [cmark-gfm](https://launchpad.net/ubuntu/+source/cmark-gfm), the version in the repository is `0.29.0.gfm.6`, but the command you're running is trying to install `0.29.0.gfm.3`.

Try updating the repository and upgrading installed packages:
```
sudo apt update
sudo apt upgrade
```

Let me know if that works or if you still have problems.

**Comment by ralf3u** [comment]
>Let me know if that helps or if you still have problems.


The answer is:

```
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 geany-plugin-preview : Depends: libcmark-gfm0 but it is not installable or
                                 libcmark-gfm0.29.0.gfm.3 but it is not installable
                        Depends: libcmark-gfm-extensions0 but it is not installable or
                                 libcmark-gfm-extensions0.29.0.gfm.3 but it is not installable
E: Unable to correct problems, you have held broken packages.
```

**Comment by xiota** [comment]
I'll double check the PPA.  Can you also check files in `/etc/apt/sources.list.d` to make sure they all refer to `lunar`?

**Comment by ralf3u** [comment]
content of the document `xiota-ubuntu-geany-plugins-lunar.list`:
```
deb https://ppa.launchpadcontent.net/xiota/geany-plugins/ubuntu/ lunar main
# deb-src https://ppa.launchpadcontent.net/xiota/geany-plugins/ubuntu/ lunar main
```

**Comment by xiota** [comment]
I'm forcing an update of the lunar package on the PPA.  As usual, takes some hours to build and update.  In case it doesn't work, can you paste the output of `apt search libcmark-gfm` ?

**Comment by ralf3u** [comment]
Does not work either.

>...output of apt search libcmark-gfm

```
Sorting... Done
Full Text Search... Done
libcmark-gfm-dev/lunar 0.29.0.gfm.6-6 amd64
  CommonMark GitHub flavor gfm library dev files

libcmark-gfm-extensions-dev/lunar 0.29.0.gfm.6-6 amd64
  CommonMark GitHub flavor gfm extensions library dev files

libcmark-gfm-extensions0.29.0.gfm.6/lunar 0.29.0.gfm.6-6 amd64
  CommonMark GitHub flavor gfm extension library

libcmark-gfm0.29.0.gfm.6/lunar 0.29.0.gfm.6-6 amd64
  CommonMark GitHub flavor gfm library

libghc-cmark-gfm-dev/lunar 0.2.5+ds1-1build2 amd64
  fast, accurate GitHub Flavored Markdown parser and renderer

libghc-cmark-gfm-doc/lunar,lunar 0.2.5+ds1-1build2 all
  fast, accurate GitHub Flavored Markdown parser and renderer; documentation

libghc-cmark-gfm-prof/lunar 0.2.5+ds1-1build2 amd64
  fast, accurate GitHub Flavored Markdown parser and renderer; profiling libraries

```

**Comment by xiota** [comment]
Pushed another update.   Let me know if it works after it rebuilds.

**Comment by ralf3u** [comment]
>Let me know if it works after it rebuilds.

It Works.

Did you try to solve the issue with the auto-sidebar?

**Comment by ralf3u** [comment]
Maybe you missed my last message, and that is why I ask again the question:

Did you try to solve the issue with the auto-sidebar?

**Comment by xiota** [comment]
Sorry, lost track of the message and forgot to respond.  The script works fine on my computer.  Which package did you install the Lua plugin from?

**Comment by ralf3u** [comment]
If I add switch-views to geanylua then this is the error-message:

![geany_auto-sidebar_2](https://github.com/xiota/geanylua-scripts/assets/68117017/fa7ca63c-e357-477e-84b7-bab409e0486d)

**Comment by xiota** [comment]
The script is trying to access document properties (filetype) before the document has been created.  Remove reference to `auto-sidebar` from `init.lua`.  I will update the scripts.

**Comment by ralf3u** [comment]
> I will update the scripts.

The scripts work now without an error-message. Thank you so much.

**Comment by ralf3u** [comment]
Is there a reason why auto-sidebar does not work any more with fountain-documents?
This is what I added:
```
  ["Fountain"] = sidebar_restore,
```

**Comment by xiota** [comment]
Is this a new install of Geany?  You have to add the Fountain filetype.  See #5.  I should probably some of add that info to the docs for the Preview plugin.

**Comment by ralf3u** [comment]
Yes! That helped me:
https://github.com/xiota/geanylua-scripts/issues/5#issuecomment-1446853464
Thanks.

**Comment by xiota** [comment]
> the fixes will be integrated in geany-plugins

Maybe someday.  It's up to the maintainers.  If you're using the plugins from my PPA, the fixes are already included.

> the focus feature

Depends on which "focus feature" you're referring to.  Focus (on a specific widget, like the preview pane) is a separate issue.  `activate` allows changing the document tab from a script.  I think in some cases, it moves focus to the editor.  I don't remember because I haven't looked at it since I fixed it.

I don't recall using `keygrab` in any of my scripts.  I just fixed it because I could.

**Comment by ralf3u** [comment]
So, the both fixes don't solve the focus-feature of the preview-pane?

**Comment by xiota** [comment]
> So, the both fixes don't solve the focus-feature of the preview-pane?

`activate` potentially moves keyboard focus to the editor.  It doesn't help with the preview pane focus issues.

**Comment by ralf3u** [comment]
>>What about the idea to fork Geanylua...
>
>Too much work to maintain a fork just for this. Ideally the existing focus function in GeanyLua would be fixed, but I have other PRs with bug fixes pending that I would want to wait for before doing anything else with it.

The last sentence sounds for me as you have an idea how to solve the focus-issue. Am I right?

**Comment by xiota** [comment]
> The last sentence sounds for me as you have an idea how to solve the focus-issue. Am I right?

No.  I don't know how to fix it.  I feel like it would be difficult, maybe not even possible.  So I don't even want to look at it, especially since I have PRs that haven't been merged yet, even though they fix significant bugs.
